### PR TITLE
Add fast paths for various `MatchData` methods.

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: src/main/ruby/truffleruby/core/regexp.rb [hunk 1]
```diff
@@ -289,10 +289,7 @@ class << self
   end
 
   def offset(idx)
-    out = []
-    out << self.begin(idx)
-    out << self.end(idx)
-    out
+    [self.begin(idx), self.end(idx)]
```
Comment (chrisseaton) [inline src/main/ruby/truffleruby/core/regexp.rb:292]: Bizarre ancient code.

File: src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java [hunk 2]
```diff
@@ -426,8 +396,24 @@ private static boolean bytesEqual(byte[] bytes, int byteLength, byte[] name, int
...
         @Specialization(guards = "inBounds(matchData, index)")
         protected Object begin(RubyMatchData matchData, int index,
+                @Cached ConditionProfile negativeBeginProfile,
+                @Cached ConditionProfile multiByteCharacterProfile,
+                @Cached RopeNodes.SingleByteOptimizableNode singleByteOptimizableNode,
                 @CachedLibrary(limit = "2") RubyStringLibrary strings) {
-            return MatchDataNodes.begin(matchData, index, strings.getRope(matchData.source));
+            // Taken from org.jruby.RubyMatchData.
+
+            final Rope matchDataSourceRope = strings.getRope(matchData.source);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:405]: Let's get the rope after the early negative check, so we don't even read it if we don't need it (same for `end`).

**Review by eregon — APPROVED** [review]
Looks good, thanks!

File: src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java [hunk 3]
```diff
@@ -259,12 +259,13 @@ protected RubyArray getIndex(RubyMatchData matchData, int index, int length,
...
                         "cachedIndex == index" })
         protected Object getIndexSymbolSingleMatch(RubyMatchData matchData, RubySymbol index, NotProvided length,
+                @Cached ConditionProfile singleBackRefProfile,
                 @Cached("index") RubySymbol cachedIndex,
                 @Cached("getRegexp(matchData)") RubyRegexp regexp,
                 @Cached("findNameEntry(regexp, index)") NameEntry name,
                 @Cached("numBackRefs(name)") int backRefs,
                 @Cached("backRefIndex(name)") int backRefIndex) {
-            if (backRefs == 1) {
+            if (singleBackRefProfile.profile(backRefs == 1)) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:268]: This doesn't need a profile, `backRefs` is `@Cached` and so PE-constant.

Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:268]: Good catch. My pattern matching was over-eager.

**Pull request merged**